web_external/util/annotationGeometryParser.js

```diff
@@ -1,154 +1,261 @@
-        this._enableState.set(id1, enabled);
+        this._trackMap.get(id1).enableState = enabled;
         return this.copy();
     }
 
+    //-------------------------------------------------------------------------
     getTrackFrameRange(id1) {
-        return this._trackRanges.get(id1);
+        return this._trackMap.get(id1).frameRange;
     }
 
+    //-------------------------------------------------------------------------
     getNextTrackId() {
-        return Math.max(...Array.from(this._trackIds)) + 1;
+        return Math.max(...this._trackMap.keys()) + 1;
     }
 
+    //-------------------------------------------------------------------------
     validateNewTrackId(trackId) {
-        return !this._trackIds.has(trackId);
+        return !this._trackMap.has(trackId);
     }
 
+    //-------------------------------------------------------------------------
     newTrack(trackId) {
-        this._trackIds.add(trackId);
-        this._trackRanges.set(trackId, [0, 0]);
-        this._enableState.set(trackId, true);
+        this._trackMap.set(trackId, new AnnotationGeometryTrack());
         return this.copy();
     }
 
+    //-------------------------------------------------------------------------
     changeTrack(trackId, newTrackId) {
         if (trackId !== newTrackId) {
-            this._trackIds.delete(trackId);
-            this._trackIds.add(newTrackId);
-            this._trackRanges.set(newTrackId, this._trackRanges.get(trackId));
-            this._enableState.set(newTrackId, this._enableState.get(trackId));
-            _.flatten(Array.from(this._frameMap.values())).forEach((geom) => {
-                if (geom.id1 !== trackId) {
-                    return;
-                }
-                geom.id1 = newTrackId;
-                if (!this._addedGeom.has(geom)) {
-                    this._editedGeom.add(geom);
+            // Reassign track in track map
+            let track = this._trackMap.get(trackId);
+            this._trackMap.set(newTrackId, track);
+            this._trackMap.delete(trackId);
+
+            // Update all geometries
+            for (let [ts0, geomSet] of this._frameMap) {
+                let geom = geomSet.get(trackId);
+                if (geom) {
+                    geom.id1 = newTrackId;
+                    geomSet.delete(trackId);
+                    geomSet.set(newTrackId, geom);
+
+                    if (!this._addedGeom.has(geom.id0)) {
+                        this._editedGeom.set(geom.id0, geom);
+                    }
                 }
-            });
+            };
+        }
+        return this.copy();
+    }
+
+    //-------------------------------------------------------------------------
+    removeTrack(trackId) {
+        for (let [ts0, geomSet] of this._frameMap) {
+            if (geomSet.has(trackId)) {
+                let state = {id1: trackId, ts0};
+                this._remove(frame, trackId, this._stateMap.get(state));
+            }
         }
```

Comment:
...you're also supposed to tell me I forgot to remove the actual *track*, here :wink:. (Which is yet another time refactoring tracks into their own type is useful; I only need to change one member, not three.)